explore/explore_internal.h

```diff
@@ -0,0 +1,347 @@
+        for (It d = pdf_first; d != pdf_last; ++d)
+          if (*d > minimum_uniform)
+            *d *= ratio;
+      }
+    }
+    
+    return S_EXPLORATION_OK;
+  }
+
+  template<typename It>
+  int enforce_minimum_probability(float minimum_uniform, bool update_zero_elements, It pdf_first, It pdf_last)
+  {
+	  typedef typename std::iterator_traits<It>::iterator_category pdf_category;
+
+	  return enforce_minimum_probability(minimum_uniform, update_zero_elements, pdf_first, pdf_last, pdf_category());
+  }
+
+  template<typename It>
+  int sample_after_normalizing(uint64_t seed, It pdf_first, It pdf_last, uint32_t& chosen_index, std::input_iterator_tag pdf_category)
+  {
+    if (pdf_first == pdf_last || pdf_last < pdf_first)
+      return E_EXPLORATION_BAD_RANGE;
+    // Create a discrete_distribution based on the returned weights. This class handles the
+    // case where the sum of the weights is < or > 1, by normalizing agains the sum.
+    float total = 0.f;
+    for (It pdf = pdf_first; pdf != pdf_last; ++pdf)
+    {
+      if (*pdf < 0)
+        *pdf = 0;
+
+      total += *pdf;
+    }
+
+    // assume the first is the best
+    if (total == 0)
+    {
+      chosen_index = 0;
+      return S_EXPLORATION_OK;
+    }
+    
+    float draw = total * uniform_random_merand48(seed);
+    if (draw > total) //make very sure that draw can not be greater than total.
+      draw = total;
+
+    float sum = 0.f;
+    uint32_t i = 0;
+    for (It pdf = pdf_first; pdf != pdf_last; ++pdf, ++i)
+    {
+      sum += *pdf;
+      if (sum > draw)
+      {
+        chosen_index = i;
+        return S_EXPLORATION_OK;
+      }
+    }
+
+    chosen_index = i - 1;
+    return S_EXPLORATION_OK;
+  }
+
+  template<typename It>
+  int sample_after_normalizing(uint64_t seed, It pdf_first, It pdf_last, uint32_t& chosen_index)
+  {
+	  typedef typename std::iterator_traits<It>::iterator_category pdf_category;
+    return sample_after_normalizing(seed, pdf_first, pdf_last, chosen_index, pdf_category());
+  }
+
+  template<typename It>
+  int sample_after_normalizing(const char* seed, It pdf_first, It pdf_last, uint32_t& chosen_index, std::random_access_iterator_tag pdf_category)
+  {
+    uint64_t seed_hash = uniform_hash(seed, strlen(seed), 0);
+    return sample_after_normalizing(seed_hash, pdf_first, pdf_last, chosen_index, pdf_category);
+  }
+
+  template<typename It>
+  int sample_after_normalizing(const char* seed, It pdf_first, It pdf_last, uint32_t& chosen_index)
+  {
+	  typedef typename std::iterator_traits<It>::iterator_category pdf_category;
+    return sample_after_normalizing(seed, pdf_first, pdf_last, chosen_index, pdf_category());
+  }
+
+  template<typename PdfIt, typename InputScoreIt, typename OutputIt>
+  int sample_after_normalizing(uint64_t seed,
+      PdfIt pdf_begin, PdfIt pdf_end, std::random_access_iterator_tag pdf_category,
+      InputScoreIt scores_begin, InputScoreIt scores_last, std::random_access_iterator_tag scores_category,
```

Comment:
Why are we using begin/end, begin/last and first/last?  It seems like we should normalize throughout the interfaces?

Reply:
done